### DISCUSSION TITLE
How to set text for each line without using replace method?

### DISCUSSION THREAD

**Discussion by nissansz** [pymupdf/PyMuPDF#2667] [General]
How to set text for each line without using replace method?

For example, I want to export all lines to a separate html, txt etc.
And then proofread each line,
then import the proofread version back to the pdf.

**Comment by nissansz** [comment]
The source pdf is something like a double-layer pdf with text over or under image.

**Reply by JorjMcKie** [reply]
Which "replace method"? Redactions?

**Comment by nissansz** [comment]
Need a method like setting text for each line.

**Reply by JorjMcKie** [reply]
There are `page.insert_text()` and `page.insert_textbox()`. Anything wrong with them?
Or use `TextWriter` which makes it easier to support different fonts.

**Comment by nissansz** [comment]
can set span directly like reading it?
span["text"] = ''AAA'?

----------------------
for i in range(0,doc.page_count):
    page = doc[i]
    for block in page.get_text("dict", flags=fitz.TEXTFLAGS_TEXT)["blocks"]:
        for line in block["lines"]:
            j=j+1
            bbox = line["bbox"]  # the line bbox
            text = ''.join([span["text"] for span in line["spans"]])  # text in line

**Reply by JorjMcKie** [reply]
Yes, but you have to somehow choose the font.
There is a 99% probability that you cannot use the original font, if you insert a different text.
If you have a font and a fontsize, you have to determine the insertion point - which is not simply the bottom-left of the bbox. I would choose the y1-value of the first span. Likewise font size. Then do this:
`page.insert_text(origin, text, font=fontname, fontsize=fontsize, overlay=True/False)`. Choose overlay as needed.

**Reply by nissansz** [reply]
what is origin? Can share a full code?

**Reply by nissansz** [reply]
what is origin? Can share a full code?

**Reply by JorjMcKie** [reply]
As I wrote:
1. You have the line bbox
2. Take `origin = line["spans"][0]["origin"]`, then `origin.x = bbox.x0`.
3. Fontsize similar `fontsize = line["spans"][0]["origin"]`line["spans"][0]["size"]`

That's all.

**Reply by nissansz** [reply]
If insert new, how to remove old?

**Reply by JorjMcKie** [reply]
Use redaction annotations.

**Reply by nissansz** [reply]
I wonder how abbbyy updates text in verify window.

**Reply by JorjMcKie** [reply]
Why not ask them 😉

**Reply by nissansz** [reply]
They are good to effect changes in verify window.
But no option to import other format to verify window.

**Reply by JorjMcKie** [reply]
As long as you don't see them writing back the changes to the original PDF, I would not be impressed at all.

**Reply by nissansz** [reply]
They can export double-layer pdf to effect the changes.

**Reply by JorjMcKie** [reply]
ok

**Reply by nissansz** [reply]
The only point is that I don't know how to export from verify window and then import proofread result.
There may be no such function.
